crates/bevy_platform_support/src/staging.rs

```diff
@@ -0,0 +1,460 @@
+        self.apply_staged_non_blocking();
+        result
+    }
+
+    /// Easily run a [`StagedRef`] function.
+    #[inline]
+    pub fn read_scope_locked<R>(&self, f: impl FnOnce(&StagedRef<T>) -> R) -> R {
+        f(&self.read_lock().as_staged_ref())
+    }
+}
+
+impl<T: StagedChanges, C: ColdStorage<T>> StagerLocked<'_, T, C> {
+    /// Allows a user to view this as a [`Stager`].
+    #[inline]
+    pub fn as_stager(&mut self) -> Stager<'_, T> {
+        Stager {
+            cold: &self.cold,
+            staged: &mut self.staged,
+        }
+    }
+
+    /// Allows a user to view this as a [`StagedRef`].
+    #[inline]
+    pub fn as_staged_ref(&self) -> StagedRef<'_, T> {
+        StagedRef {
+            cold: &self.cold,
+            staged: &self.staged,
+        }
+    }
+}
+
+impl<T: StagedChanges, C: ColdStorage<T>> StagedRefLocked<'_, T, C> {
+    /// Allows a user to view this as a [`StagedRef`].
+    #[inline]
+    pub fn as_staged_ref(&self) -> StagedRef<'_, T> {
+        StagedRef {
+            cold: &self.cold,
+            staged: &self.staged,
+        }
+    }
+}
+
+impl<'a, T: StagedChanges> Clone for StagedRef<'a, T> {
+    fn clone(&self) -> Self {
+        Self {
+            staged: self.staged,
+            cold: self.cold,
+        }
+    }
+}
+
+impl<T: StagedChanges> Default for AtomicStageOnWrite<T>
+where
+    T::Cold: Default,
+{
+    fn default() -> Self {
+        Self::new(T::Cold::default())
+    }
+}
+
+impl<T: StagedChanges> ColdStorage<T> for RwLockReadGuard<'_, T::Cold> {}
+
+impl<T: StagedChanges> ColdStorage<T> for &'_ T::Cold {}
+
+#[cfg(test)]
+mod tests {
+    use crate::{collections::HashMap, prelude::Vec};
+
+    use super::*;
+
+    #[derive(Default)]
+    struct StagedNumVec {
+        added: Vec<u32>,
+        changed: HashMap<usize, u32>,
+    }
+
+    impl StagedChanges for StagedNumVec {
+        type Cold = Vec<u32>;
+
+        fn apply_staged(&mut self, storage: &mut Self::Cold) {
+            storage.append(&mut self.added);
+            for (index, new) in self.changed.drain() {
+                storage[index] = new;
+            }
+        }
+
+        fn any_staged(&self) -> bool {
+            !self.added.is_empty() || !self.changed.is_empty()
+        }
+    }
+
+    // This is commented out, as it intentionally does not compile. This demonstrates how `AtomicStageOnWrite` prevents deadlock using the borrow checker.
+    // #[test]
+    // fn test_no_compile_for_deadlock() {
+    //     let mut stage_on_write = AtomicStageOnWrite::<StagedNumVec>::default();
+    //     let reading = stage_on_write.read_lock();
+    //     stage_on_write.apply_staged_non_blocking();
+    // }
+
+    #[test]
+    fn test_simple_stage() {
+        let mut data = StageOnWrite::<StagedNumVec>::default();
+        data.stage_scope_locked(|stager| stager.staged.added.push(5));
+        let full = data.apply_staged_for_full();
+        assert_eq!(full[0], 5);
```

Comment:
```suggestion
        assert_eq!(full[..], &[5]);
```
Stronger test just to ensure the value isn't added twice, etc.